File: MtxVec.Delphi/List_Func.pas
unit List_Func;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Grids, StdCtrls, ExtCtrls;

type
  TfrmListFunc = class(TForm)
    StringGrid1: TStringGrid;
    Panel1: TPanel;
    Label1: TLabel;
    Button1: TButton;
    Button2: TButton;
    procedure FormCreate(Sender: TObject);
    procedure FormResize(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
  private
    { Private declarations }
    LastIndex: Integer;
    LastSearchText: String;
    LastFindAll: boolean;
    function ListToStringGrid(AFileName: String): Integer;
    function DoSearch(FindText: String; FindAllWords: boolean): Integer;
  public
    { Public declarations }
  end;

var
  frmListFunc: TfrmListFunc;

implementation

Uses Find_Form;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

function TfrmListFunc.DoSearch(FindText: String; FindAllWords: boolean): Integer;

var SearchText : Array[1..100] of String;
    NumWords   : Integer;

  function FoundTextIn(S: String): boolean;
  var i: Integer;
  begin
    Result := False;
    S := UpperCase(S);

    if FindAllWords then
    begin
      for i:=1 to NumWords do
         if Pos(SearchText[i],S) = 0 then exit;
     result := true;
    end else
    begin
      for i:=1 to NumWords do
        if Pos(SearchText[i],S)>0 then
        begin
          result := true;
           exit;
        end;
    end;
  end;

  Function SearchInLines(Lines:TStrings):Boolean;
  var i : Integer;
  begin
    result:=False;
    for i:=0 to Lines.Count-1 do
      if FoundTextIn(Lines[i]) then
      begin
        result:=True;
        break;
      end;
  end;

  Procedure SplitWords(S: String);
  var i : Integer;
  begin
    NumWords:=0;
    Repeat
      i:=Pos(' ',S);
      if i>0 then
      begin
        Inc(NumWords);
        SearchText[NumWords]:=UpperCase(Copy(S,1,i-1));
        Delete(S,1,i);
      end;
    Until (i=0) or (S='');
    if S<>'' then
    begin
      Inc(NumWords);
      SearchText[NumWords]:=UpperCase(S);
    end;
  end;

var tmpText: String;
    i: Integer;
    FirstIndex: Integer;
begin
  Result := -1;
  tmpText := FindText;
  SplitWords(tmpText);
  FirstIndex := LastIndex+1;
  for i := FirstIndex to StringGrid1.RowCount-1 do
    if SearchInLines(StringGrid1.Rows[i]) then
    begin
      Result := i;
      StringGrid1.Row := i;
      break;
    end;
end;


procedure TfrmListFunc.FormCreate(Sender: TObject);
var Total : Integer;
begin
  Total := ListToStringGrid('List_Functions.txt');
  Label1.Caption := 'Function list: Count = '+IntToStr(Total) + '  (not counting overloaded functions)';
  LastIndex := -1;
  LastSearchText := '';
  LastFindAll := false;
end;

function TfrmListFunc.ListToStringGrid(AFileName: String): Integer;
var i,j, Len: Integer;
    StrList: TStringList;
begin
  Result := 0;
  With StringGrid1 do
  begin
    RowCount :=1;
    Cells[0,0] := '(Unit) Name';
    Cells[1,0] := 'Description';
  end;
  If FileExists(AFileName) then
  begin
    StrList := TStringList.Create;
    try
      StrList.LoadFromFile(AFileName);
      for i := 0 to StrList.Count-1 do
      begin
          StringGrid1.RowCount := i+1;
          Len := Length(StrList[i]);
          j:= Pos(Chr(9),StrList[i]);
          StringGrid1.Cells[0,i] := Copy(StrList[i],0,j-1);
          StringGrid1.Cells[1,i] := Copy(StrList[i],j+1,len-j);
          if StringGrid1.Cells[1,i]<>'' then Result := Result +1;
      end;
      StringGrid1.FixedRows := 1;
    finally
      StrList.Free;
    end;
  end;
end;

procedure TfrmListFunc.FormResize(Sender: TObject);
begin
  StringGrid1.ColWidths[1] := StringGrid1.Width - StringGrid1.ColWidths[0];
end;

procedure TfrmListFunc.Button1Click(Sender: TObject);
begin
  if LastIndex <> -1 then LastIndex := DoSearch(LastSearchText,LastFindAll)
  else
  with TfrmFind.Create(Self) do
  try
    ShowModal
  finally
    if ModalResult = mrOk	then
    begin
      LastIndex := DoSearch(FindText,FindAll);
      LastSearchText := FindText;
      LastFindAll := FindAll;
    end;
    Free;
  end;
  if LastIndex <> -1 then Button1.Caption := 'Next'
  else Button1.Caption := 'Find';
end;

procedure TfrmListFunc.Button2Click(Sender: TObject);
begin
    LastIndex := -1;
    LastSearchText := '';
    LastFindAll := false;
    Button1.Caption := 'Find';
end;

initialization
  RegisterClass(TfrmListFunc);

end.

File: MtxVec.CPP/List_Func.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "List_Func.h"
#include "Find_Form.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmListFunc *frmListFunc;
//---------------------------------------------------------------------------
__fastcall TfrmListFunc::TfrmListFunc(TComponent* Owner)
        : TForm(Owner)
{
  int Total;
  AnsiString FilePath;
  FilePath = ExtractFilePath(Application->ExeName);

  Total = ListToStringGrid(FilePath + "List_Functions.txt");
  Label1->Caption = "Function list: Count = " +IntToStr(Total) + "  (not counting overloaded functions)";
  LastIndex = -1;
  LastSearchText = "";
  LastFindAll = false;
}
//---------------------------------------------------------------------------
  bool FoundTextIn(TStrings *SearchText, AnsiString S, bool FindAllWords)
  {
    int i;
    bool Result = false;
    S = S.UpperCase();

    if (FindAllWords)  {

      for (i = 0; i < SearchText->Count;i ++)  {
         if (S.AnsiPos(SearchText->Strings[i]) == 0) return Result;
      }
      Result = true;

    } else {

      for (i = 0; i < SearchText->Count; i++) {

        if (S.AnsiPos(SearchText->Strings[i]) > 0)  {

           Result = true;
           return Result;
        }
      }
   }

   return Result;
  }

  bool SearchInLines(TStrings* SearchText, TStrings *Lines, bool FindAllWords) {

    bool result = false;
    int i;

    for (i = 0; i < Lines->Count; i++) {

      if (FoundTextIn(SearchText, Lines->Strings[i], FindAllWords)) {

        result = true;
        return result;
      }
    }
    return result;
  }

  void  SplitWords(TStringList *SearchText, AnsiString S)
  {
    int i;
    while ((i != 0) && (S != ""))  {

      i = S.AnsiPos(" ");
      if (i > 0) {

        SearchText->Add( (S.SubString(1,i-1)).UpperCase());
        S = S.Delete(1,i);
      }
    }
    if (S != "") {

      SearchText->Add(S.UpperCase());
    }
  }

//---------------------------------------------------------------------------
int __fastcall TfrmListFunc::DoSearch(AnsiString FindText, bool FindAllWords)
{
  TStringList *SearchText;
  int Result = -1;
  int  i, FirstIndex;

  SearchText = new TStringList();
  try
     {
        SplitWords(SearchText, FindText);
        FirstIndex = LastIndex+1;
        for (i = FirstIndex; i < StringGrid1->RowCount; i++) {
          if (SearchInLines(SearchText, StringGrid1->Rows[i], FindAllWords)) {

            Result = i;
            StringGrid1->Row = i;
            break;
          }
        }
     }
  __finally
  {
       delete SearchText;
  }

  return Result;
}

//---------------------------------------------------------------------------
int __fastcall TfrmListFunc::ListToStringGrid(AnsiString AFileName)
{
  int i,j, Len;
  TStringList *StrList;
  int Result = 0;

  StringGrid1->RowCount = 1;
  StringGrid1->Cells[0][0] = "(Unit) Name";
  StringGrid1->Cells[1][0] = "Description";

  if (FileExists(AFileName)) {

    StrList = new TStringList();
    try
      {
      StrList->LoadFromFile(AFileName);
      for (i = 0; i< StrList->Count; i++) {

          StringGrid1->RowCount = i+1;
          Len = StrList->Strings[i].Length();
          j = StrList->Strings[i].Pos("\t");
          StringGrid1->Cells[0][i] = StrList->Strings[i].SubString(0,j-1);
          StringGrid1->Cells[1][i] = StrList->Strings[i].SubString(j+1,Len-j);
          if (StringGrid1->Cells[1][i] != "") Result++;
      }
      StringGrid1->FixedRows = 1;
      }
    __finally {

        delete StrList;
    }
  }

  return Result;
}
//---------------------------------------------------------------------------
void __fastcall TfrmListFunc::Button1Click(TObject *Sender)
{
  TfrmFind *FindForm;

  if (LastIndex != -1) { LastIndex = DoSearch(LastSearchText,LastFindAll); }  else
  {
     FindForm = new TfrmFind(this);
     try       {

        FindForm->ShowModal();
        
        }
     __finally {

          if (FindForm->ModalResult == mrOk) {

              LastIndex = DoSearch(FindForm->FindText,FindForm->FindAll);
              LastSearchText = FindForm->FindText;
              LastFindAll = FindForm->FindAll;
          }
          delete FindForm;
     }
  }
  if (LastIndex != -1) {Button1->Caption = "Next";}
  else Button1->Caption = "Find";
}
//---------------------------------------------------------------------------
void __fastcall TfrmListFunc::Button2Click(TObject *Sender)
{
  LastIndex = -1;
  LastSearchText = "";
  LastFindAll = false;
  Button1->Caption = "Find";
}
//---------------------------------------------------------------------------
void __fastcall TfrmListFunc::FormResize(TObject *Sender)
{
  StringGrid1->ColWidths[1] = StringGrid1->Width - StringGrid1->ColWidths[0];
}
//---------------------------------------------------------------------------

